pint_server/app.py

```diff
@@ -306,70 +452,178 @@ def _get_azure_images_for_region_state(region, state):
+        )
+    elif provider == 'microsoft':
+        images = _get_azure_deprecatedby_images_for_region(deprecatedby,
+                                                           region)
+    elif hasattr(PROVIDER_IMAGES_MODEL_MAP[provider], 'region'):
+        images = PROVIDER_IMAGES_MODEL_MAP[provider].query.filter(
+            PROVIDER_IMAGES_MODEL_MAP[provider].region == region,
+            PROVIDER_IMAGES_MODEL_MAP[provider].state == ImageState.deprecated,
+            PROVIDER_IMAGES_MODEL_MAP[provider].deprecatedon < deprecatedby,
+        )
+
+    return images
+
+
+def get_provider_images_to_be_deletedby(deletedby, provider, region=None):
+    images = query_deletedby_images_in_provider_region(
+        deletedby, provider, region)
+
+    extra_attrs = {}
+    if region and provider == 'microsoft':
+        extra_attrs['region'] = region
+
+    return formatted_provider_images(provider, images,
+                                     extra_attrs=extra_attrs)
+
+
+def _query_image_in_azure_region(image_name, region):
+    # lookup environment for given region, assuming unique per region
+    environment_name = _get_azure_environment_name_for_region(region)
+
+    # retrieve matching images for region
+    images = MicrosoftImagesModel.query.filter(
+        MicrosoftImagesModel.environment == environment_name,
+        MicrosoftImagesModel.name == image_name)
+
+    return images
+
+
+def query_image_in_provider_region(image_name, provider, region=None):
+    if region is None:
+        images = PROVIDER_IMAGES_MODEL_MAP[provider].query.filter(
+            PROVIDER_IMAGES_MODEL_MAP[provider].name == image_name)
+    # microsoft needs special handling for region queries
+    elif provider == 'microsoft':
+        images = _query_image_in_azure_region(image_name, region)
+
+    # if provider has regions retrieve matching images for region
+    elif (hasattr(PROVIDER_IMAGES_MODEL_MAP[provider], 'region')):
+        images = PROVIDER_IMAGES_MODEL_MAP[provider].query.filter(
+            PROVIDER_IMAGES_MODEL_MAP[provider].region == region,
+            PROVIDER_IMAGES_MODEL_MAP[provider].name == image_name)
+
+    return images
+
+
+def get_image_deletiondate_in_provider(image, provider, region=None):
+
+    images = query_image_in_provider_region(image, provider, region)
+
+    # depending on provider there may be multiple images all of
+    # which should have the same state, deprecatedon and deletedon
+    # values. Use a set to eliminate duplicates, and a namedtuple
+    # helper to create a hashable object.
+    image_set = {DeletionDetails(i.state,
+                                 i.deprecatedon,
+                                 i.deletedon)
+                 for i in images
+                 if i.state in [ImageState.deprecated,
+                                ImageState.deleted]}
+
+    # TODO(rtamalin): is this even possible?
+    if len(image_set) > 1:
```

Comment:
Well that'll depend on the client of the code, if `image` is a full image name than `> 1` cannot happen (assuming that the __hash__ method of a namedtuple produces consistently the same value for the same content). But if the content of `image` is such that multiple matches can occurr then the set could be longer than 1.

Reply:
So right now the `query_image_in_provider_region()` call is performing an exact match search for the provided image name matching the image name in the provider's images table, so we will only ever see results for one image, and all of the results should be for that image name.

My question was more about whether it would ever be possible for the deletion related attrs used in constructing the set values, namely the `state`, `deprecatedon` and `deleted`, to be different for the same image, at least intentionally? Basically, for a provider with region specific images, for a specific image name, could we intentionally see different states, deletedon or deprecatedon values for that image?

If not then we can get rid of this check ;-)

Reply:
In any given provider the set should always be consistent, i.e. if `suse-sles-sap-12-sp3-v20220304-hvm-ssd-x86_64` is set to deprecate in one region of the provider it should be set to the same state (deprecated) in all regions of that provider.

Reply:
FYI due to issues seen with microsoft (Blackforest environment) and AWS (some regions have had images already deprecated) we can have more than one set of results in the image_set, so we handle this by reporting deletiondate based upon the earliest possible deprecatedon or deletedon date depending on whether the image is in the deprecated or deleted state.